dirty_memory_manager.hh

```diff
@@ -28,12 +28,21 @@ class sstable_write_permit final {
     }
 
 public:
-    sstable_write_permit(sstable_write_permit&&) noexcept = default;
-    sstable_write_permit& operator=(sstable_write_permit&&) noexcept = default;
+    sstable_write_permit(sstable_write_permit&& o) noexcept : _permit(std::exchange(o._permit, std::nullopt)) {}
```

Comment:
The "=default" doesn't work any more? Or what's the motivation here?

Reply:
I'll add the motivation to the commit log.

The default move implementation calls std::optional move operator which moves the value contents but leaves the source optional engaged if it was already engaged (just containing a moved-from value).
This makes it harder to check for implementing the bool operator below, and be used for `has_sstable_write_permit`

Reply:
In general, after a move one should not use the moved-from value. Maybe the caller should have used std::exchange() if it intended to call operator bool later?

Reply:
I think we can move the logic up a layer to `flush_permit` below.
so it can contain std::optional<sstable_write_permit> and can use std::exchange to release it and has_value() to check its presence.